include/aos/common/pkcs11/pkcs11.hpp

```diff
@@ -0,0 +1,718 @@
+    /**
+     * pkcs11 object handle.
+     */
+    ObjectHandle GetHandle() const { return mHandle; }
+
+private:
+    ObjectHandle mHandle;
+};
+
+/**
+ * A helper class for a base Cryptoki API.
+ */
+class Utils {
+public:
+    /**
+     * Creates an object instance.
+     *
+     * @param session session context.
+     * @param allocator allocator for token/session objects.
+     */
+    Utils(pkcs11::SessionContext& session, Allocator& allocator)
+    {
+        (void)session;
+        (void)allocator;
+    }
+
+    /**
+     * Creates an RSA key pair on the token.
+     *
+     * @param id token id.
+     * @param label label.
+     * @param bitsCount key size.
+     * @return RetWithError<SharedPtr<PrivateKey>>.
+     */
+    RetWithError<SharedPtr<PrivateKey>> GenerateRSAKeyPairWithLabel(
+        const Array<uint8_t>& id, const String& label, size_t bitsCount)
+    {
+        (void)id;
+        (void)label;
+        (void)bitsCount;
+
+        return {nullptr, Error::Enum::eNone};
+    }
+
+    /**
+     * Creates a ECDSA key pair on the token using curve.
+     *
+     * @param id token id.
+     * @param label label.
+     * @param curve elliptic curve.
+     * @return RetWithError<SharedPtr<PrivateKey>>.
+     */
+    RetWithError<SharedPtr<PrivateKey>> GenerateECDSAKeyPairWithLabel(
+        const Array<uint8_t>& id, const String& label, EllipticCurve curve)
+    {
+        (void)id;
+        (void)label;
+        (void)curve;
+
+        return {nullptr, Error::Enum::eNone};
+    }
+
+    /**
+     * Imports a certificate onto the token.
+     *
+     * @param id certificate id.
+     * @param label certificate label.
+     * @param cert certificate.
+     * @return Error.
+     */
+    Error ImportCertificate(const Array<uint8_t>& id, const String& label, const crypto::x509::Certificate& cert)
+    {
+        (void)id;
+        (void)label;
+        (void)cert;
+
+        return {};
+    }
+
+    /**
+     * Destroys a previously imported certificate.
+     *
+     * @param id certificate id.
+     * @param label certificate label.
+     * @return Error.
+     */
+    Error DeleteCertificate(const Array<uint8_t>& id, const String& label)
+    {
+        (void)id;
+        (void)label;
+
+        return {};
+    }
+
+    /**
+     * Checks whether certificate with specified attributes has been imported.
+     *
+     * @param issuer certificate issuer.
+     * @param serialNumber certificate serialNumber.
+     * @return Error.
+     */
+    RetWithError<bool> HasCertificate(const Array<uint8_t>& issuer, const Array<uint8_t>& serialNumber)
+    {
+        (void)issuer;
+        (void)serialNumber;
+
+        return {true};
+    }
+
+    /**
+     * Retrieves a previously created asymmetric key pair.
+     *
+     * @param id key id.
+     * @param label key label.
+     * @return RetWithError<SharedPtr<PrivateKey>>.
+     */
+    RetWithError<SharedPtr<PrivateKey>> FindPrivateKey(const Array<uint8_t>& id, const String& label)
```

Comment:
What would happen if utils and session will be destroyed? Will this key works as expected?

Reply:
Yes, it is.  SharedPtr is allocated with external allocator. Produced handles are not bound to one session.